### PR TITLE
DTR replication support

### PR DESCRIPTION
**Pull request by gregsidelinger**
Support for Docker Trusted Registry, https://docs.docker.com/ee/dtr/.

I've tested this on DTR 2.7.3. Push and pull support are both working.

### CODE REVIEW

**Comment by xaleeks** [comment]
Thanks for the PR, as this is integration with a specific version of DTR hosted registry, we need to consider testing and support for future versions as well before merging.  But it's great that you got this replication working.

**Comment by gregsidelinger** [comment]
Im not aware of any version specific stuff I did but will do a test with a much older version of DTR we still have setup later this week. I can also stand up a 2.6.x version to test without much trouble.  As for future compatibility that would need to be addressed when DTR adds a v1 api, breaks v0 or changes how UCP and DTR work together since they are tied at the hips right now which I have no insight into.

**Comment by gregsidelinger** [comment]
I was able to test the patch against DTR 2.4.13 today. Both pushing and pulling replication worked without any issues.  It was able to make organizations, repositories and then sync the images without any issues while pushing.  The pulling worked without any issues.

**Comment by bitsf** [comment]
@gregsidelinger need fix code format as ci failed

**Comment by gregsidelinger** [comment]
Wow, fixing three log lines cause almost all CI tests to fail with new errors.  Looking into why go vet is angry, may need to rebase as it has been a month.

**Comment by gregsidelinger** [comment]
I reproduced the new CI errors locally after merging with master. I updated my code with the new factory that all replication adaptors now implement and go vet is clean.  Will run a few tests tomorrow when I have access to DTR and update the PR.

**Comment by gregsidelinger** [comment]
I cleaned up the formatting of the code to make all of the CI checks happy and have all the commits signed.  Let me know if there is anything else.

File: src/replication/adapter/dtr/client.go [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+	}
+	return result, nil
+}
+
+// getNamespaces returns DTR namespaces.  DTR also calles these orgs and accounts depending on where you look
+func (c *Client) getNamespaces() ([]Account, error) {
+	var accounts []Account
+	var response Accounts
+
+	endpoint := fmt.Sprintf("%s/enzi/v0/accounts?limit=100", c.url)
```
Comment (bitsf) [inline src/replication/adapter/dtr/client.go]: Can you share the api document that you are referenced here

Comment (gregsidelinger) [reply src/replication/adapter/dtr/client.go]: https://docs.docker.com/v17.09/datacenter/dtr/2.3/reference/api/

Unfortunately this API is no longer documented in the latest API docs however it is still there.  It is the only way to create orgs as the core DTR api does not support any of those features based on how tied it is to UCP.

**Comment by gregsidelinger** [comment]
> Do you have an existing env that we can have a test

All of my current DTR environments are within my companies network which I can not open up.  I think a limited version of UCP with DTR can be stood up without license keys.  Or they do offer a 30 day trial key.

https://docs.docker.com/ee/ucp/admin/install/
https://docs.docker.com/ee/dtr/admin/install/

**Comment by bitsf** [comment]
trial license is not enough for testing, we can't add it in CI.
@xaleeks Can we or CNCF bought a DTR license, or ask for a long term trial license from docker team ?

**Comment by gregsidelinger** [comment]
> trial license is not enough for testing, we can't add it in CI.
> @xaleeks Can we or CNCF bought a DTR license, or ask for a long term trial license from docker team ?

FYI, Mirantis purchased Docker Enterprise's business last month and DTR was part of that so Docker Inc no longer owns DTR.

**Comment by MnrGreg** [comment]
@gregsidelinger @bitsf @xaleeks 

I'm sure there are a number of organizations that would benefit from this Registry Provider integration- myself included. Seems like quite a heavy lift though to get the licensing, environment and CI tests working.

Would enabling this provider through some kind of alpha feature flag be an option in the interim?

**Comment by xaleeks** [comment]
let's discuss this further since I have some concerns on upgrades and migrations.  Can you ping me xalex@vmware.com ?  @gregsidelinger

**Comment by MnrGreg** [comment]
@gregsidelinger @xaleeks wondering if there has been any movement here?

Is enabling this through some kind of pre-alpha feature flag an option in the interim?

**Comment by gregsidelinger** [comment]
@MnrGreg I sent an email to @xaleeks a few weeks back and have not heard back yet.  Have you tried doing a build from the PR to verify it works for you?

@xaleeks if a --alpha flag of some type would get this merged in let me know and I can look into adding it.

**Comment by MnrGreg** [comment]
@gregsidelinger yes, we have a build running with this PR and the replication with DTR is working well.

Example of pull replication from a DTR hosted alpine container below:

```shell
2020-01-31T08:24:19Z [DEBUG] [/replication/adapter/dtr/adapter.go:50]: Entering DTR newAdapter
2020-01-31T08:24:19Z [DEBUG] [/replication/adapter/dtr/adapter.go:54]: Setting up DTR for basic auth
2020-01-31T08:24:19Z [DEBUG] [/replication/adapter/dtr/adapter.go:54]: Setting up DTR for basic auth
2020-01-31T08:24:19Z [DEBUG] [/common/utils/registry/auth/tokenauthorizer.go:214]: scopes parsed from request: repository:system/alpine:pull
...
2020-01-31T08:28:32Z [DEBUG] [/common/utils/registry/auth/tokenauthorizer.go:214]: scopes parsed from request: repository:system/alpine:pull,push
2020-01-31T08:28:32Z [DEBUG] [/common/utils/registry/auth/tokenauthorizer.go:239]: get token for scope repository:system/alpine:pull,push from cache
2020-01-31T08:28:33Z [DEBUG] [/common/utils/registry/transport.go:51]: 201 | PUT http://goharbor-harbor-core/v2/system/alpine/manifests/3.9
```

```shell
10.1.0.32 - - [31/Jan/2020:08:24:26 +0000] "HEAD /v2/system/alpine/manifests/3.4 HTTP/1.1" 404 93 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:24:26 +0000] "HEAD /v2/system/alpine/blobs/sha256:b7c5ffe56db790f91296bcebc5158280933712ee2fc8e6dc7d6c96dbb1632431 HTTP/1.1" 404 157 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:24:26 +0000] "POST /v2/system/alpine/blobs/uploads/ HTTP/1.1" 202 0 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:24:26 +0000] "PUT /v2/system/alpine/blobs/uploads/ea265b15-cacc-456c-afa4-bddadfebd799?_state=M1iM2VdIl8W_qlFWpsOea4lfwGZWaAJU65la8M0sr5d7Ik5hbWUiOiJzeXN0ZW0vYWxwaW5lIiwiVVVJRCI6ImVhMjY1YjE1LWNhY2MtNDU2Yy1hZmE0LWJkZGFkZmViZDc5OSIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAyMC0wMS0zMVQwODoyNDoyNi42ODEyMzM3NVoifQ%3D%3D&digest=sha256:b7c5ffe56db790f91296bcebc5158280933712ee2fc8e6dc7d6c96dbb1632431 HTTP/1.1" 201 0 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:24:26 +0000] "HEAD /v2/system/alpine/blobs/sha256:c1e54eec4b5786500c19795d1fc604aa7302aee307edfe0554a5c07108b77d48 HTTP/1.1" 404 157 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:24:27 +0000] "POST /v2/system/alpine/blobs/uploads/ HTTP/1.1" 202 0 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:24:27 +0000] "PUT /v2/system/alpine/blobs/uploads/ac0d74bf-1a22-4ca1-9c66-cc4d3bd94dd3?_state=g-mZenec0VrCRpmyQZ1TORBYlCot9gfDze9VQ6XpPoN7Ik5hbWUiOiJzeXN0ZW0vYWxwaW5lIiwiVVVJRCI6ImFjMGQ3NGJmLTFhMjItNGNhMS05YzY2LWNjNGQzYmQ5NGRkMyIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAyMC0wMS0zMVQwODoyNDoyNy41MzIzMjkwNDJaIn0%3D&digest=sha256:c1e54eec4b5786500c19795d1fc604aa7302aee307edfe0554a5c07108b77d48 HTTP/1.1" 201 0 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:26:31 +0000] "PUT /v2/system/alpine/manifests/3.4 HTTP/1.1" 201 0 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:26:31 +0000] "HEAD /v2/system/alpine/manifests/3.4 HTTP/1.1" 200 528 "" "harbor-registry-client"
10.1.0.32 - - [31/Jan/2020:08:26:32 +0000] "HEAD /v2/system/alpine/manifests/3.9 HTTP/1.1" 404 93 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:26:32 +0000] "HEAD /v2/system/alpine/blobs/sha256:055936d3920576da37aa9bc460d70c5f212028bda1c08c0879aedf03d7a66ea1 HTTP/1.1" 404 157 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:26:32 +0000] "POST /v2/system/alpine/blobs/uploads/ HTTP/1.1" 202 0 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:26:32 +0000] "PUT /v2/system/alpine/blobs/uploads/e9766336-2c3a-4274-beea-999ba8b39f4a?_state=1KG5J0ttyX1Ng2oZNSfsTHLKm2_VdidXPH8Md5oSk2h7Ik5hbWUiOiJzeXN0ZW0vYWxwaW5lIiwiVVVJRCI6ImU5NzY2MzM2LTJjM2EtNDI3NC1iZWVhLTk5OWJhOGIzOWY0YSIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAyMC0wMS0zMVQwODoyNjozMi45Njg0NjQ5NjJaIn0%3D&digest=sha256:055936d3920576da37aa9bc460d70c5f212028bda1c08c0879aedf03d7a66ea1 HTTP/1.1" 201 0 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:26:32 +0000] "HEAD /v2/system/alpine/blobs/sha256:e7c96db7181be991f19a9fb6975cdbbd73c65f4a2681348e63a141a2192a5f10 HTTP/1.1" 404 157 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:26:34 +0000] "POST /v2/system/alpine/blobs/uploads/ HTTP/1.1" 202 0 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:26:56 +0000] "PUT /v2/system/alpine/blobs/uploads/8883699b-b256-4044-8a64-6e51cadc6e5b?_state=wKbgGsM7xqnOAGPPqV78E7si2GYlEc2GCRwJVeZTA5t7Ik5hbWUiOiJzeXN0ZW0vYWxwaW5lIiwiVVVJRCI6Ijg4ODM2OTliLWIyNTYtNDA0NC04YTY0LTZlNTFjYWRjNmU1YiIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAyMC0wMS0zMVQwODoyNjozNC4zNjQ0NTM4NzZaIn0%3D&digest=sha256:e7c96db7181be991f19a9fb6975cdbbd73c65f4a2681348e63a141a2192a5f10 HTTP/1.1" 201 0 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:28:33 +0000] "PUT /v2/system/alpine/manifests/3.9 HTTP/1.1" 201 0 "" "harbor-replication-service"
10.1.0.32 - - [31/Jan/2020:08:28:33 +0000] "HEAD /v2/system/alpine/manifests/3.9 HTTP/1.1" 200 528 "" "harbor-registry-client"
```

@xaleeks what are your thoughts around catering for this adapter. Would an interim feature flag be possible?

**Comment by bitsf** [comment]
Please add some more UT cases in order to make the coverage not go down too much.

**Comment by gregsidelinger** [comment]
@bitsf  sounds good, I will rebase and add some UI tests some time this week.

**Comment by gregsidelinger** [comment]
Can anyone point me to an example of a UI test for an existing replication adaptor?  I poked around and could not make sense of where Karma was loading tests from.

**Comment by bitsf** [comment]
Add some Backend UT is enough now, we haven't add UI UT for each adaptors too.

**Comment by gregsidelinger** [comment]
I added a few UT for the backend client.   I will try and get to a few more in a couple of days to get the coverage above 30%.

**Comment by gregsidelinger** [comment]
I added a couple more tests and the coverage is at 59%.  Let me know what else is needed.

**Comment by MnrGreg** [comment]
Hey folks, wondering what’s happening with this PR? Anything preventing the merge?

@steven-zou @ywk253100

**Comment by bjethwan** [comment]
@gregsidelinger 
I am very much interested in this PR.
I can sure give a hand, testing this within my company with a few DTR/UCP instances.
Meanwhile, I was thinking if Go lang has a way to dynamically load this adapter at the run time (i.e. something similar to dynamic jar loading in Java)? So the harbor team can support that dynamic loading, that way it will be up to the end-user to pick and choose the adapters.

**Comment by dajose** [comment]
@gregsidelinger I see this PR no longer matches with the latest master :( could you update it?

**Comment by gregsidelinger** [comment]
Looks like some of the imports need updated.  My current diff is below.  However master is currently not compiling for me based on some other things currently broken in master.  I will commit my changes once the master branch is fixed and I can run the test suite.

```
diff --git a/src/replication/adapter/dtr/adapter.go b/src/replication/adapter/dtr/adapter.go
index 7d4758978..722038255 100644
--- a/src/replication/adapter/dtr/adapter.go
+++ b/src/replication/adapter/dtr/adapter.go
@@ -7,8 +7,8 @@ import (
        "strings"

        "github.com/goharbor/harbor/src/common/utils"
-       "github.com/goharbor/harbor/src/common/utils/log"
-       "github.com/goharbor/harbor/src/common/utils/registry/auth"
+       "github.com/goharbor/harbor/src/lib/log"
+       "github.com/goharbor/harbor/src/pkg/registry/auth"
        adp "github.com/goharbor/harbor/src/replication/adapter"
        "github.com/goharbor/harbor/src/replication/adapter/native"
        "github.com/goharbor/harbor/src/replication/model"
diff --git a/src/replication/adapter/dtr/client.go b/src/replication/adapter/dtr/client.go
index cd547e5c9..fb12dd527 100644
--- a/src/replication/adapter/dtr/client.go
+++ b/src/replication/adapter/dtr/client.go
@@ -12,7 +12,7 @@ import (
        "strings"

        common_http "github.com/goharbor/harbor/src/common/http"
-       "github.com/goharbor/harbor/src/common/utils/log"
+       "github.com/goharbor/harbor/src/lib/log"
        adp "github.com/goharbor/harbor/src/replication/adapter"
        "github.com/goharbor/harbor/src/replication/model"
        "github.com/goharbor/harbor/src/replication/util"
```

**Comment by steven-zou** [comment]
@gregsidelinger 

Recently, we're trying to set up a compliance verification testing framework for all the replication adapters. This testing framework will help us to verify and monitor if the adapter is still working well (as we know, the registries the adapters are targeting are also making changes sometimes).

The purpose of setting up that framework is to make sure we do not deliver broken adapters to the community or finding the broken adapters because of changes in targeting registries immediately.

After that, we'll continue to accept the new replication adapters to the supporting list. It will be soon (several weeks). Any progress, I'll let you know. Please stay tuned!

**Comment by gregsidelinger** [comment]
I have an updated PR for the changes in 2.0.  I still need to do some testing however the test suite is passing. I use helm to deploy harbor so I'm not yet running 2.0 since the chart still needs updated based on the 2.0.0 release notes.  Once the chart is updated I do some manual testing and commit the changes.

Let me know what needs done for the compliance verification testing framework once more of that is finalized.

**Comment by wy65701436** [comment]
We will consider to merge it after the testing framework is ready. @bitsf , please give update about the progress of testing framework development, and the document of how to use it.

**Comment by gregsidelinger** [comment]
@bitsf any update on the testing framework?

**Comment by MnrGreg** [comment]
Lots of bug fixes and new features in v2.1.0 but sad to see this one didn't make it. Community, lets up vote this for the new release!?

@bitsf @wy65701436 where can we track the Testing Framework?

**Comment by steven-zou** [comment]
> @bitsf any update on the testing framework?

@gregsidelinger 

Thanks for the patience.

We have created a verification framework to test and verify if the related replication adapters are still working well with Harbor, see the repo https://github.com/goharbor/replication-verification.

Please follow the readme to see if your adapter can pass the testing and then paste the testing results here. If cases are passed, we can accept and merge this adapter that's expected for a long while.

Any help needed, please ping @bitsf / me. Thanks.

**Comment by bitsf** [comment]
@gregsidelinger because we don't have the environment, so you just need to upload the test result to prove everything is working fine.
you can create a test case file like `master/cases/Quay.robot`, and prepare some images in your DTR registry, and run the test job with `python -u -m robot.run ...`, of course you need prepare the env reference the CI file `.github/workflows/main.yml`

**Comment by gregsidelinger** [comment]
I'm a bit confused about the Post actions as documented in [CONTRIBUTING](https://github.com/goharbor/replication-verification/edit/master/CONTRIBUTING.md).  For steps 2-4.  Am I suppose to send you the PR for my test case and include login info for a DTR setup?

```
## 3. Post actions

1. Attach the gif record generated at the step 4 of section 2 to the adapter code PR mentioned at step 4 of section 1
as a proof of the PR.
2. Provide the AK info that we can add to the github Secrets.
3. Change .github/workflows/main.yml#112 and your case to using the AK in secrets.
4. Raise PR to submit the testing cases for the new adapter to this repository for future regular executions.
```

The DTR I have access to is sitting inside my companies network and is not exposed to the Internet.  I can setup and run the test case once but there is no way I can expose DTR on the Internet let along provide login access to it.

**Comment by steven-zou** [comment]
> The DTR I have access to is sitting inside my companies network and is not exposed to the Internet. I can setup and run the test case once but there is no way I can expose DTR on the Internet let along provide login access to it.

@gregsidelinger 

Got the point. You can use your DTR env to run the case offline and provide the results here. But it's internal env and then can not be used in the regular replication adapter verification pipeline.

Let's skip the post actions so far.

After PR is merged, we can continue to think about how to activate the verification test cases in the framework later.

cc @bitsf

**Review by bitsf — APPROVED** [review]

**Comment by bitsf** [comment]
@gregsidelinger please rebase commit and merge

**Comment by bitsf** [comment]
@gregsidelinger hi, it always has conflicts and can't merge.
you should submit only 1 commit that would make the commit history looks better.
1. git checkout goharbor/master
2. git merge [commitid] --no-commit
3. git commit
4. git push -f gregsidelinger dtr

**Pull request merged**